lib/class-wp-rest-shortcodes-controller.php

```diff
@@ -0,0 +1,196 @@
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_shortcode_output( $request ) {
+		global $post;
+		global $wp_embed;
+		$style         = '';
+		$js            = '';
+		$type          = 'html';
+		$output        = '';
+		$yt_pattern    = '#https?://(?:www\.)?(?:youtube\.com/watch|youtu\.be/)#';
+		$vimeo_pattern = '#https?://(.+\.)?vimeo\.com/.*#';
+		$args          = $request->get_params();
+		$post          = isset( $args['postId'] ) ? get_post( $args['postId'] ) : null;
+		$shortcode     = isset( $args['shortcode'] ) ? trim( $args['shortcode'] ) : '';
+		$cache_key     = 'shortcode_' . md5( serialize( $args ) );
+		$data          = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			return rest_ensure_response( $data );
+		}
+		// Initialize $data.
+		$data = array(
+			'html'  => $output,
+			'type'  => $type,
+			'style' => $style,
+			'js'    => $js,
+		);
+
+		if ( empty( $shortcode ) ) {
+			$data['html'] = __( 'Enter something to preview', 'gutenberg' );
+			return rest_ensure_response( $data );
+		}
+
+		if ( ! empty( $post ) ) {
+			setup_postdata( $post );
+		}
+
+		// Since the [embed] shortcode needs to be run earlier than other shortcodes.
+		if ( has_shortcode( $shortcode, 'embed' ) ) {
+			$output = $wp_embed->run_shortcode( $shortcode );
+		} else {
+			$output = do_shortcode( $shortcode );
+		}
+
+		if ( empty( $output ) ) {
+			$data['html'] = __( 'Sorry, couldn\'t render a preview', 'gutenberg' );
+			return rest_ensure_response( $data );
+		}
+
+		// Check if shortcode is returning a video. The video type will be used by the frontend to maintain 16:9 aspect ratio.
+		// TODO: Extend embed video compare to other services too, such as videopress.
+		if ( has_shortcode( $shortcode, 'video' ) ) {
+			$type = 'video';
+		} elseif ( has_shortcode( $shortcode, 'embed' ) && preg_match( $yt_pattern, $shortcode ) ) {
+			$type = 'video';
+		} elseif ( has_shortcode( $shortcode, 'embed' ) && preg_match( $vimeo_pattern, $shortcode ) ) {
+			$type = 'video';
+		} else {
+			$type = 'html';
+			// Gallery and caption shortcodes need the theme style to be embedded in the shortcode preview iframe.
+			if ( has_shortcode( $shortcode, 'gallery' ) || has_shortcode( $shortcode, 'caption' ) || has_shortcode( $shortcode, 'wp_caption' ) ) {
+				$style = '<link rel="stylesheet" type="text/css" href="' . get_stylesheet_uri() . '" />';
+			}
+
+			// Playlist shortcodes need the playlist JS to be embedded in the shortcode preview iframe.
+			if ( has_shortcode( $shortcode, 'playlist' ) ) {
+				ob_start();
+				wp_footer();
+				$js = ob_get_clean();
+			}
+		}
+
+		$data = array(
+			'html'  => $output,
+			'type'  => $type,
+			'style' => $style,
+			'js'    => $js,
+		);
+
+		// Caches the result for 12 hours.
+		set_transient( $cache_key, $data, 12 * HOUR_IN_SECONDS );
+
+		return rest_ensure_response( $data );
```

Comment:
This function seems a bit complex, I'm not confident in reviewing this code personally, might be good if someone else could take a look but can I ask how did you come up with the logic here and all the specific cases? don't we have a generic way to render all shortcodes?

Reply:
Here's what I've done:
- The crux of this API is exactly what you've mentioned - it makes use of `do_shortcode()` as a generic way to render all shortcodes. There are a few issues and edge cases that I've additionally handled.
- `[embed]` shortcode being a "special" one that needs to be run before all other shortcodes, I've implemented `run_shortcode()` ( the comments in this <a href="https://core.trac.wordpress.org/browser/tags/4.9/src/wp-includes/class-wp-embed.php#L48"> core source file </a> describes this point).
- `postId` is needed since some shortcodes such as `[gallery]` expect the context (reference <a href="https://github.com/WordPress/gutenberg/issues/1054#issuecomment-308670584">comment</a> and gallery shortcode performing a `get_post` in the <a href="https://core.trac.wordpress.org/browser/trunk/src/wp-includes/media.php#L1678">core source file</a>.). So I'm sending the postId to the REST API to set the global post variable.
- When `[playlist]` is run through `do_shortcode()`, a script element is returned in the response that contains a data structure. This shortcode needs additional styles and scripts to act on the data structure. Following this shortcode in the <a href="https://github.com/WordPress/WordPress/blob/master/wp-includes/media.php#L1955">core source files</a>, I landed upon the need to include `wp_footer()` in the response sent by the API. Pl check this <a href="https://github.com/WordPress/gutenberg/pull/4710#commitcomment-27107497">comment discussion</a> too.
- `[gallery]` too had its own quirk of needing styles to be returned in the response. 
- So, the function determines if the shortcode is a gallery or playlist to return the appropriate styles and scripts.
- Implements caching via transients. Though this might be redundant with the new `withAPIData` change which handles caching on the front end.
- Finally, I determine if the `type` of the shortcode is video. Since videos need a 16:9 aspect ratio for responsive render (reference <a href="https://github.com/WordPress/gutenberg/blob/master/components/sandbox/index.js#L86">here</a>).

Happy to provide more clarity if needed.

Reply:
In fact, I think that all of this function could be replaced with something that does the following:

```php
global $post;
$post = get_post( $args['postId'] );
setup_postdata();
ob_start();
?>
<!DOCTYPE html>
<html>
	<head>
		<?php wp_head(); ?>
	</head>
	<body>
		<?php echo apply_filters( 'the_content', $args['shortcode'] ); ?>
		<?php wp_footer(); ?>
	</body>
</html>
<?php
$output = ob_get_clean();
```

This entire output could then be passed into the `Sandbox` component, though I realize that the component is expecting the supplied `html` to be in the `body`. This could reduce the special cases needing to be accounted for.

Reply:
Thanks for reviewing and providing your feedback @westonruter . I will implement your suggestion and test it thoroughly.

Reply:
@westonruter I just took a closer look at this method : 
- The Sandbox component (which is also used by the embed block) expects the `type` parameter (if content is a video or otherwise), based on which it <a href="https://github.com/WordPress/gutenberg/blob/master/components/sandbox/index.js#L85">explicitly sets the height</a> based on a 16:9 ratio for responsive content.

 The method you've outlined doesn't determine type of content as video or otherwise. I will check if there's another way to determine this, else I feel this method might not work out. Will be happy to hear your take on this too.

Reply:
I altered the rest endpoint to only return 'html' type and this seems to work with embed or shortcode

Reply:
It's also not using this rest endpoint (maybe why it has not affected the embed)